File: Kotlin/PrimeSpeedTest/src/main/kotlin/PrimeFinder.kt
import java.time.LocalDateTime
import kotlin.math.sqrt

class PrimeFinder {
    private var primes = mutableListOf<ULong>(2u, 3u, 5u)
    private var sqrtBase : ULong = 2u

    fun calculateFirstNPrimes(n: Int) {
        if (primes.size >= n) return

        var current = getCurrent()

        while (primes.size < n) {
            checkIfPrime(current - 1u)
            checkIfPrime(current + 1u)

            current += 6u
        }

        if (primes.size > n){
            primes.removeAt(primes.size - 1)
        }
    }

    private fun setSqrtBase(n: ULong) {
        if (n - sqrtBase * sqrtBase > sqrtBase + sqrtBase)
        {
            sqrtBase++;
        }
    }

    private fun getCurrent() : ULong {
        var current = primes.last()
        current += 6u - current % 6u

        if (current - 1u == primes.last())
        {
            checkIfPrime(current + 1u)
            current += 6u
        }

        return current
    }

    private fun checkIfPrime(n: ULong) {
        setSqrtBase(n)

        for (i in (2..primes.size)){
            if (n % primes[i] == 0.toULong()) {
                return
            }

            if (primes[i] > sqrtBase) {
                break
            }
        }

        primes.add(n)
    }

    fun getPrimesCount() : Int {
        return primes.size
    }
}

File: C++/PrimeSpeedTest/PrimeSpeedTest.Console/PrimeFinder.cpp
#include "PrimeFinder.h"

void Primes::prime_finder::set_sqrt_base(const long number)
{
	if (number - sqrt_base_ * sqrt_base_ > sqrt_base_ + sqrt_base_)
	{
		sqrt_base_++;
	}
}

void Primes::prime_finder::check_if_prime(const long number)
{
	set_sqrt_base(number);

	for (unsigned long long i = 2; i < primes_.size(); i++)
	{
		if (number % primes_[i] == 0)
		{
			return;
		}

		if (primes_[i] > sqrt_base_)
		{
			break;
		}
	}

	primes_.push_back(number);
}

long Primes::prime_finder::get_current()
{
	long current = primes_.back();
	current += 6 - current % 6;

	if (current - 1 == primes_.back())
	{
		check_if_prime(current + 1);
		current += 6;
	}

	return current;
}

Primes::prime_finder::prime_finder()
{
	sqrt_base_ = 2;
	primes_.push_back(2);
	primes_.push_back(3);
}

void Primes::prime_finder::calculate_first_n_primes(const int n)
{
	if (primes_.size() >= n)
	{
		return;
	}

	long current = get_current();

	while (primes_.size() < n)
	{
		check_if_prime(current - 1);
		check_if_prime(current + 1);

		current += 6;
	}

	if (primes_.size() > n)
	{
		primes_.pop_back();
	}
}

unsigned long long Primes::prime_finder::get_calculated_primes_length() const
{
	return primes_.size();
}

long Primes::prime_finder::get_last_prime() const
{
	return primes_.back();
}
